pkg/kubelet/images/image_manager.go

```diff
@@ -177,3 +231,57 @@ func applyDefaultImageTag(image string) (string, error) {
 	}
 	return image, nil
 }
+
+// isEnsuredBySecret - returns true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image/
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (pulledBySecret, ensuredBySecret bool) {
+	if imageRef == "" {
+		return
+	}
+	if ensureSecretPulledImage[imageRef] != nil {
+		pulledBySecret = true
+	}
+
+	img := image.Image
+	repoToPull, _, _, err := parsers.ParseImageName(img)
+	if err != nil {
+		return
+	}
+
+	if m.keyring == nil {
+		return
+	}
+	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, *m.keyring)
+	if err != nil {
+		return
+	}
+
+	creds, withCredentials := keyring.Lookup(repoToPull)
+	if !withCredentials {
+		return
+	}
+
+	for _, currentCreds := range creds {
+		auth := &runtimeapi.AuthConfig{
+			Username:      currentCreds.Username,
+			Password:      currentCreds.Password,
+			Auth:          currentCreds.Auth,
+			ServerAddress: currentCreds.ServerAddress,
+			IdentityToken: currentCreds.IdentityToken,
+			RegistryToken: currentCreds.RegistryToken,
+		}
+
+		hash := kubecontainer.HashAuth(auth)
```

Comment:
Would this need to check whether the provided credentials were extracted from the imagePullSecrets and not from another docker auth provder in the keyring?

Reply:
> Wondering if this need to check whether the provided credentials were extracted from the imagePullSecrets and not from another docker auth provder in the keyring?

see: https://github.com/kubernetes/kubernetes/pull/94899/files#diff-7a6e7cc9b15ce528a5503356cbe743bcR31-L59

Reply:
Right, so was wondering if we account for when the auth config came from the default keyring `m.keyring` and not from image pull secrets.

https://github.com/kubernetes/kubernetes/blob/f4112710f5bbfbd9cc7a68ede0df519af30d5e40/pkg/kubelet/kuberuntime/kuberuntime_image.go#L38-L41

Here it seems like we hash all the provided auth configs, not just image pull secrets. Is that expected?

Reply:
ah.. Yes the idea is to match the hash of the auth cred passed to the container runtime that was used when the pull succeeded... to what is about to become a subsequent attempt. If a cred in the creds I get from the keyring already passed authentication, then it should pass this time as well was my thought. I was thinking we could time stamp it and force authentication with the same creds after some duration.  Or I could just use a single boolean for the imageRef indicating it was pulled with auth and thus should be pulled each time to re-verify, but that would be less performant for the PullIfNotPresent (with auth) scenario.

Reply:
Pinged you on slack maybe we can have a talk about this. Near as I can tell it was coded this way since the first code drop July of 2016. @Random-Liu /  @feiskyer do you remember or know of any expectation here wrt. @andrewsykim 's question?

Reply:
I'm not sure what we'd change based on accounting for that tbh? (and if anything started changing based on that it'd likely be a pretty big conceptual change?)

Reply:
> 

nod..

Reply:
> 

nod

Reply:
> I'm not sure what we'd change based on accounting for that tbh? (and if anything started changing based on that it'd likely be a pretty big conceptual change?)

nod..